src/backend/cdb/dispatcher/cdbdisp.c

```diff
@@ -540,7 +554,10 @@ AtAbort_DispatcherState(void)
 	 * Cleanup all outbound dispatcher states belong to
 	 * current resource owner and its children
 	 */
-	CdbResourceOwnerWalker(CurrentResourceOwner, cdbdisp_cleanupDispatcherHandle);
+	if (ERRCODE_TO_CATEGORY(elog_geterrcode()) != ERRCODE_INSUFFICIENT_RESOURCES)
+		CdbResourceOwnerWalker(CurrentResourceOwner, cdbdisp_cleanupDispatcherHandle);
+	else
+		CdbResourceOwnerWalker(CurrentResourceOwner, cdbdisp_destroyDispatcherHandle);
```

Comment:
`pgindent` your new changes or maybe separate the callback into a separate variable that is set by the condition above

Reply:
or
```suggestion
	CdbResourceOwnerWalker(CurrentResourceOwner,
	ERRCODE_TO_CATEGORY(elog_geterrcode()) != ERRCODE_INSUFFICIENT_RESOURCES ?
	cdbdisp_cleanupDispatcherHandle :
	cdbdisp_destroyDispatcherHandle);
```